proof/invariant-abstract/AInvs.thy

```diff
@@ -22,26 +22,32 @@ lemma pred_tcb_at_upd_apply:
   pred_tcb_at proj P t (s\<lparr>kheap := (kheap s)(t := p' t)\<rparr>)"
   by (simp add: pred_tcb_at_def obj_at_def)
 
-lemma thread_set_tcb_arch_is_schedulable_bool[wp]:
-  "\<lbrace>\<lambda>s. is_schedulable_bool (cur_thread s) s\<rbrace>
-     thread_set (\<lambda>tcb. tcb\<lparr>tcb_arch := arch_tcb_context_set us (tcb_arch tcb)\<rparr>) t
-   \<lbrace>\<lambda>rv s. is_schedulable_bool (cur_thread s) s\<rbrace>"
+lemma thread_set_tcb_arch_ct_schedulable[wp]:
+  "thread_set (\<lambda>tcb. tcb\<lparr>tcb_arch := arch_tcb_context_set us (tcb_arch tcb)\<rparr>) t \<lbrace>ct_schedulable\<rbrace>"
   apply (simp add: thread_set_def)
   apply (rule hoare_seq_ext[OF _ assert_get_tcb_ko'])
   apply (wpsimp wp: set_object_wp)
-  apply (fastforce simp: is_schedulable_bool_def is_sc_active_def get_tcb_def ko_atD
+  apply (fastforce simp: schedulable_def is_sc_active_def get_tcb_def ko_atD
                          in_release_queue_def
                   split: option.splits )
   done
 
+lemma thread_set_tcb_arch_ct_not_running[wp]:
+  "thread_set (\<lambda>tcb. tcb\<lparr>tcb_arch := arch_tcb_context_set us (tcb_arch tcb)\<rparr>) t \<lbrace>\<lambda>s. \<not> ct_running s\<rbrace>"
+  apply (simp add: thread_set_def)
+  apply (rule hoare_seq_ext[OF _ assert_get_tcb_ko'])
+  apply (wpsimp wp: set_object_wp)
+  apply (clarsimp simp: ct_in_state_def pred_tcb_at_def obj_at_def split: if_splits)
+  done
+
 text \<open>The top-level invariance\<close>
 
 lemma akernel_invs:
   "\<lbrace>\<lambda>s. invs s \<and> (e \<noteq> Interrupt \<longrightarrow> ct_running s) \<and>
         scheduler_action s = resume_cur_thread \<and>
-        is_schedulable_bool (cur_thread s) s\<rbrace>
-     (call_kernel e)
-   \<lbrace>\<lambda>rv s. (invs s \<and> (ct_running s \<or> ct_idle s))\<rbrace>"
+        (ct_running s \<longrightarrow> ct_schedulable s)\<rbrace>
+   call_kernel e
+   \<lbrace>\<lambda>_ s. (invs s \<and> (ct_running s \<or> ct_idle s))\<rbrace>"
```

Comment:
Feel free to ignore this nitpick, but the postcondition doesn't need the outer brackets